src/Behat/Config/Formatter/Formatter.php

```diff
@@ -55,8 +78,63 @@ public function toArray(): array
         return $this->settings;
     }
 
+    public function toPhpExpr(): Expr
+    {
+        $formatterObject =  $this->builderFactory->new(new FullyQualified(self::class));
+
+        $expr = $this->applyBaseSettings($formatterObject);
+
+        if ($this->settings === []) {
+            $args = $this->builderFactory->args([$this->name]);
+        } else {
+            $args = $this->builderFactory->args([$this->name, $this->settings]);
+        }
+        $formatterObject->args = $args;
+
+        return $expr;
+    }
+
     public static function defaults(): array
     {
         return [];
     }
+
+    public function applyBaseSettings(Expr $expr): Expr
+    {
+        foreach ($this->settings as $settingName => $setting) {
+            $functionName = Formatter::FORMATTER_FUNCTION_NAMES_PER_SETTING[$settingName] ?? null;
+            if ($functionName !== null) {
+                $args = $this->builderFactory->args([$setting]);
+                $expr = $this->builderFactory->methodCall($expr, $functionName, $args);
+                unset($this->settings[$settingName]);
+            }
+        }
+        return $expr;
+    }
+
+    protected function toPhpExprForNamedFormatter(): Expr
+    {
+        $formatterObject =  $this->builderFactory->new(new FullyQualified(static::class));
+
+        $expr = $this->applyBaseSettings($formatterObject);
+
+        $defaults = static::defaults();
+        $argValues = [];
+        foreach ($defaults as $name => $defaultValue) {
```

Comment:
We look through each of the default values for the formatter and only if the value passed does not match the default, we set the corresponding parameter in the constructor of the formatter